src/session.ts

```diff
@@ -22,36 +30,116 @@ export type State =
   setTxSigning(): void;
   setTxPending(tx: TransactionResponse): void;
   setTxConfirmed(tx: TransactionReceipt): void;
   setChangedAccount(address: string): void;
 }
+let walletConnect: any;
+
+const modalStore = writable<ConnectionModalState>({ status: ConnectionModalStateType.Close, modalProps: null });
+export const store = derived(modalStore, ($store) => $store);
 
 export const loadState = (initial: State): Store => {
   const store = writable<State>(initial);
   const session = window.localStorage.getItem("session");
-
+  function newWalletConnect(config: Config): WalletConnect {
+
+    walletConnect = new WalletConnect({
+      bridge: config.radicleBridge.bridge,
+      qrcodeModal: {
+        open: (uri: string, onClose, _opts?: unknown) => {
+          modalStore.set({ status: ConnectionModalStateType.Open, modalProps: { uri, config } });
+          console.log(_opts);
+          onClose();
+        },
+        close: () => {
+          modalStore.set({ status: ConnectionModalStateType.Close, modalProps: null });
+        },
+      },
+    });
+
+    return walletConnect;
+  }
   if (session) store.set({ connection: Connection.Connected, session: JSON.parse(session) });
 
+  let network: any;
+  let provider: any;
+  let signer: any;
+
+  // Connect to a wallet using walletconnect
+  const connectWalletConnect = async (config: Config) => {
+    walletConnect = newWalletConnect(config);
+    provider = new ethers.providers.Web3Provider(window.ethereum);
+    signer = new WalletConnectSigner(walletConnect, provider);
+
+    //Todo : check wallet state in the store before attempting to connect
+    const state = get(store);
+    const session = window.localStorage.getItem("session");
+
+    if (session && walletConnect.connected) {
+      store.set({ connection: Connection.Connected, session: JSON.parse(session) });
+      return;
+    }
+    assertEq(state.connection, Connection.Disconnected);
+    store.set({ connection: Connection.Connecting });
+
+    try {
+      await walletConnect.connect();
+      const address = await signer.getAddress();
+      const tokenBalance: BigNumber = await config.token.balanceOf(address);
+      const session = { address, tokenBalance, tx: null };
+      const provNetwork = await ethers.providers.getNetwork(
+        signer.walletConnect.chainId
+      );
+      network = {
+        name: provNetwork.name,
+        chainId: provNetwork.chainId,
+      };
+      if (network.chainId !== config.network.chainId) {
+        walletConnect.killSession();
+        throw `Network mismatch`;
+      }
+      config = new Config(network, provider, signer);
+
+      store.set({ connection: Connection.Connected, session });
+      saveSession({ ...session, tokenBalance: null });
+
+    } catch (e) {
+      console.log(e);
+      assertEq(state.connection, Connection.Disconnected);
+      store.set({ connection: Connection.Disconnected });
+      assert(e, "Could not connect to wallet connect");
```

Comment:
What's going on here, these 3 lines of code don't really make sense..

Reply:
it's line 129 that doesnt make sense.. although I've removed it

Basically, when walletConnect initializes it sets the connection state to connecting, so If there's an error we should set the connection state back to disconnect

Reply:
Yeah 129 didn't make sense. But 131 is also strange, are you trying to throw an error? `e` is always going to be true here so there's nothing to assert. You should just do `throw new Error("Could not connect to wallet")`

Reply:
oh yeah. you mentioned the assert method just checks for values and returns true or false